pkg/cassandra/auth.go

```diff
@@ -13,7 +14,11 @@ import (
 const JmxInitContainer = "jmx-credentials"
 
 // ApplyAuth modifies the dc config depending on whether auth is enabled in the cluster or not.
-func ApplyAuth(dcConfig *DatacenterConfig, authEnabled bool) {
+func ApplyAuth(dcConfig *DatacenterConfig, authEnabled bool) error {
+
+	if dcConfig.PodTemplateSpec == nil {
+		return errors.New("PodTemplateSpec was nil, cannot add auth settings")
```

Comment:
I'd be curious to learn what others think, but I'm torn on the idea of returning errors when `PodTemplateSpec` is nil, for a few reasons:

1. Such errors are not the user's fault but rather the developer's, and yet users would see these messages in logs, but how would they be able to fix them by themselves?
2. Such errors being the developer's fault – our fault :-) – I'd rather see them panic here, so that our unit tests can "fail spectacularly" as you say, and thus quickly detect such problems. If the issue stems from setting auth to false, maybe we'd need to improve the existing tests so that they can catch this one, and other similar nil pointer issues. But returning an error without fixing the issue would just make the controller enter an infinite loop of failed reconcile attempts.
3. Speaking of which, the real fix is in datacenters.go, and your changes look good there. IMHO the changes in that file are enough to consider the problem solved. In all places where `dcConfig.PodTemplateSpec` is read, I'd simply remove the nil checks and consider that this field is assumed to never be nil.

Reply:
I'm 👎 on returning an error in functions that work with PodTemplateSpec. I agree about the real fix being in datacenters.go and that the changes look good. 

I also agree that this is something that should be handled with test coverage vs making functions unnecessarily convoluted. If we aren't confident in our ability to ensure PodTemplateSpec will be initialized before it's use and having adequate test coverage, then maybe we shouldn't export `dcConfig.PodTemplateSpec`, i.e., change it to `dcConfig.podTemplateSpec` and provide a method on `dcConfig` for accessing it. That method would ensure it is non-nil. I'm good though with the changes in datacenters.go :)

Reply:
RE errors being returned: I think that's some good feedback. The invariant we are trying to maintain is that the PodTemplateSpec must never be nil. I have done this by returning a runtime error, and it is correct to say that (because of the way we handle such errors) it probably isn't quite the optimal design.

The question is how we can maintain this invariant via a compile time check, perhaps the solution is to make the PodTemplateSpec a value type instead of a pointer? That way it can never be nil and we can remove the nil checks everywhere.

The reality is that (while it may have originally made sense for the PodTemplateSpec to be a pointer as it was optional) the logic has evolved and it is no longer optional at all. We should perhaps reflect that in the types.

Reply:
For the record, we had a similar situation with `CassandraConfig` and its fields, which were all pointers until a while ago; and we had nil checks everywhere. They were subsequently turned into value types and that indeed made things much cleaner. The 2 aspects to watch when doing this change imho are: turning nilness tests into equality tests; and how things get serialized/deserialized.